src/index.js

```diff
@@ -1,35 +1,76 @@
 const nodePath = require('path');
 
+export const defaultDropExtensions = [
+  '.spec',
+  '.test',
+  '.story',
+  '.stories',
+];
+
+const normalisePathSep = path => (path
+  ? path.replace(/\\/g, '/')
+  : path
+);
+const removeLeadingDotSlash = path => (path && path.match(/^\.\//)
+  ? path.replace(/^\.\//, '')
+  : path
+);
+const buildPluginConfig = opts => ({
+  dropAllFilenames: opts && opts.dropAllFilenames,
+  dropExtensions: (opts && opts.dropExtensions) || defaultDropExtensions,
+  separator: opts && opts.separator,
+  userSetRoot: opts && removeLeadingDotSlash(normalisePathSep(opts.root)),
+});
+const getConfigIfNodeTransformable = (path, state) => {
+  // Node matches default placeholder (added in v1).
+  if (path.node.name === '__filenamespace') {
+    return buildPluginConfig(state.opts);
+  }
+  // Node matches custom placeholder (added in v2).
+  // Note: always choose 1st one if duplicates exist.
+  const placeholder = (state.opts.customPlaceholders || [])
+    .filter(x => x && x.placeholder === path.node.name)[0];
+  if (placeholder) {
+    return buildPluginConfig(placeholder);
+  }
+  // Not a transformable node.
+  return false;
+};
+
 export default ({ types: t }) => ({
   visitor: {
     Identifier(path, state) {
-      if (path.node.name === '__filenamespace') {
-        // Get user settings
-        const { separator, root, dropAllFilenames } = state.opts;
+      const pluginConfig = getConfigIfNodeTransformable(path, state);
+      if (pluginConfig) {
+        // Plugin config
+        const {
+          dropAllFilenames,
+          dropExtensions,
+          separator,
+          userSetRoot,
+        } = pluginConfig;
 
         // Get file paths
-        const { filename, sourceRoot } = state.file.opts;
+        const cwd = normalisePathSep(state.cwd);
+        const sourceRoot = normalisePathSep(state.file.opts.sourceRoot);
+        const filename = normalisePathSep(state.file.opts.filename);
         const dirname = nodePath.dirname(filename);
-        const basename = nodePath.basename(filename, nodePath.extname(filename));
+        const basename = [nodePath.extname(filename)].concat(dropExtensions).reduce(
+          (acc, ext) => nodePath.basename(acc, ext),
+          filename,
+        );
 
         // Get namespace root directory (defaults to projectRoot)
         // modified via user specified plugin option "root"
-        let userSetRoot = root;
-        let namespaceRoot = sourceRoot || state.cwd;
+        let namespaceRoot = sourceRoot || cwd;
```

Comment:
sourceRoot seems to cause path to be removed in some instances like tests in mocha